plugin/evm/acp176/acp176.go

```diff
@@ -0,0 +1,197 @@
+//
+// BaseFee = MinBaseFee * e^(Excess / (Target() * TargetToPriceUpdateConversion))
+func (s *State) BaseFee() *big.Int {
+	target := s.Target()
+	priceUpdateConversion, err := safemath.Mul(TargetToPriceUpdateConversion, target)
+	if err != nil {
+		priceUpdateConversion = math.MaxUint64
+	}
+
+	bigExcess := new(big.Int).SetUint64(uint64(s.Gas.Excess))
+	bigPriceUpdateConversion := new(big.Int).SetUint64(uint64(priceUpdateConversion))
+	return fakeExponential(minBaseFee, bigExcess, bigPriceUpdateConversion)
+}
+
+// AdvanceTime increases the gas capacity and decreases the gas excess based on
+// the elapsed seconds.
+func (s *State) AdvanceTime(seconds uint64) {
+	targetPerSecond := s.Target()
+	maxPerSecond, err := safemath.Mul(TargetToMax, targetPerSecond)
+	if err != nil {
+		maxPerSecond = math.MaxUint64
+	}
+	maxCapacity, err := safemath.Mul(TimeToFillCapacity, maxPerSecond)
+	if err != nil {
+		maxCapacity = math.MaxUint64
+	}
+	s.Gas = s.Gas.AdvanceTime(
+		maxCapacity,
+		maxPerSecond,
+		targetPerSecond,
+		seconds,
+	)
+}
+
+// ConsumeGas decreases the gas capacity and increases the gas excess by
+// gasUsed + extraGasUsed. If the gas capacity is insufficient, an error is
+// returned.
+func (s *State) ConsumeGas(
+	gasUsed uint64,
+	extraGasUsed *big.Int,
+) error {
+	var err error
+	s.Gas, err = s.Gas.ConsumeGas(gas.Gas(gasUsed))
+	if err != nil {
+		return err
+	}
+	if extraGasUsed != nil {
+		if !extraGasUsed.IsUint64() {
+			return fmt.Errorf("%w: extraGasUsed (%d) exceeds MaxUint64",
+				gas.ErrInsufficientCapacity,
+				extraGasUsed,
+			)
+		}
+		s.Gas, err = s.Gas.ConsumeGas(gas.Gas(extraGasUsed.Uint64()))
+	}
+	return err
+}
+
+// UpdateTargetExcess updates the targetExcess to be as close as possible to the
+// desiredTargetExcess without exceeding the maximum targetExcess change.
+func (s *State) UpdateTargetExcess(desiredTargetExcess gas.Gas) {
+	previousTargetPerSecond := s.Target()
+	s.TargetExcess = targetExcess(s.TargetExcess, desiredTargetExcess)
+	newTargetPerSecond := s.Target()
+	s.Gas.Excess = scaleExcess(
+		s.Gas.Excess,
+		newTargetPerSecond,
+		previousTargetPerSecond,
+	)
+}
+
+// DesiredTargetExcess calculates the optimal desiredTargetExcess given the
+// desired target.
+//
+// This could be solved directly by calculating D * ln(desiredTarget / P) using
+// floating point math. However, it introduces inaccuracies. So, we use a binary
+// search to find the closest integer solution.
+func DesiredTargetExcess(desiredTarget gas.Gas) gas.Gas {
+	return gas.Gas(sort.Search(maxTargetExcess, func(targetExcessGuess int) bool {
+		state := State{
+			TargetExcess: gas.Gas(targetExcessGuess),
+		}
+		return state.Target() >= desiredTarget
+	}))
+}
+
+// fakeExponential approximates factor * e ** (numerator / denominator) using
+// Taylor expansion:
+// e^x = 1 + x/1! + x^2/2! + x^3/3! + ...
+func fakeExponential(factor, numerator, denominator *big.Int) *big.Int {
```

Comment:
nit: perhaps rename this to approximateExponential, since it's not totally fake either? 🤔

Reply:
This function is copied from geth, I don't think we should change the name.